### PR TITLE
CipherSuitePolicy implementation

### PR DESCRIPTION
**Pull request by krwq**

### CODE REVIEW

File: src/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.OpenSslName.Lookup.Linux.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (krwq) [inline src/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.OpenSslName.Lookup.Linux.cs]: remove

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c [hunk 2]
```diff
@@ -560,46 +319,91 @@ CryptoNative_SslCtxSetCertVerifyCallback(SSL_CTX* ctx, SslCtxSetCertVerifyCallba
...
+            (tlsver[0] == '.' && tlsver[1] >= '0' && tlsver[1] <= '2' && tlsver[2] == 0);
+    }
+    else
     {
-        // No minimum security policy, same as OpenSSL 1.0
-        SSL_CTX_set_security_level(ctx, 0);
+        // if we don't know it assume it is new
+        // worst case scenario OpenSSL will ignore it
+        *isTls12OrLower =
+            strncmp("DTLSv", version, 5) == 0 ||
```
Comment (bartonjs) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: There could be another version of DTLS for 1.3 (https://datatracker.ietf.org/doc/draft-ietf-tls-dtls13/), so this should be more defensive.

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c [hunk 3]
```diff
@@ -560,46 +319,91 @@ CryptoNative_SslCtxSetCertVerifyCallback(SSL_CTX* ctx, SslCtxSetCertVerifyCallba
...
+            strncmp("SSLv", version, 4) == 0;
     }
 
-    return SSL_CTX_set_cipher_list(ctx, cipherString);
+    return ret;
 }
 
+int32_t CryptoNative_Tls13Supported()
+{
+    return API_EXISTS(SSL_CTX_set_ciphersuites);
```
Comment (bartonjs) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: This seems like it currently wouldn't build with non-portable against 1.1.0 or 1.0.x...

File: src/System.Net.Security/src/System.Net.Security.csproj [hunk 4]
```diff
@@ -142,6 +142,7 @@
   </ItemGroup>
   <ItemGroup Condition=" '$(TargetsWindows)' == 'true'">
     <Compile Include="System\Net\CertificateValidationPal.Windows.cs" />
+	<Compile Include="System\Net\Security\CipherSuitesPolicy.Windows.cs" />
```
Comment (bartonjs) [inline src/System.Net.Security/src/System.Net.Security.csproj]: NIt: spacing

File: src/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.Linux.cs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                }
+
+                using (SafeSslHandle ssl = SafeSslHandle.Create(innerContext, false))
+                {
+                    if (ssl.IsInvalid)
+                    {
+                        throw OpenSsl.CreateSslException(SR.net_allocate_ssl_context_failed);
+                    }
+
+                    using (var tls13cipherSuites = new OpenSslStringBuilder())
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.Linux.cs]: Nit: `tls13cipherSuites` => `tls13CipherSuites`

File: src/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.Linux.cs [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+using System.Diagnostics;
+using System.IO;
+using System.Security.Authentication;
+using System.Text;
+using Ssl = Interop.Ssl;
+using OpenSsl = Interop.OpenSsl;
+
+namespace System.Net.Security
+{
+    public sealed class CipherSuitesPolicy
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.Linux.cs]: Since this type is public, it should be broken up into the public portion and a PAL portion.  The public portion should call into the PAL, but shouldn't have any platform-specific code in it.

File: src/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.Linux.cs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+            switch (policy)
+            {
+                case EncryptionPolicy.RequireEncryption:
+                    return RequireEncryptionDefault;
+                case EncryptionPolicy.AllowNoEncryption:
+                    return AllowNoEncryptionDefault;
+                case EncryptionPolicy.NoEncryption:
+                    return NoEncryptionDefault;
+                default:
+                    Debug.Assert(false, "This should not be possible");
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.Linux.cs]:
```suggestion
                    Debug.Fail($"Unknown EncryptionPolicy value ({policy})");
```

File: src/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.OpenSslName.Lookup.Linux.cs (hunk 1 above)
Comment (bartonjs) [reply src/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.OpenSslName.Lookup.Linux.cs]: I'm not sure what the "remove" is for.  If it's the file, OK.  If it's the license header, no... it should still be around.

File: src/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.OpenSslName.Lookup.Linux.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+using Microsoft.Win32.SafeHandles;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Authentication;
+using System.Security.Cryptography;
+
+namespace System.Net.Security
+{
+    using CS = TlsCipherSuite;
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.OpenSslName.Lookup.Linux.cs]: That using alias doesn't seem to be buying much...

File: src/System.Net.Security/tests/FunctionalTests/SslStreamNegotiatedCipherSuiteTest.cs [hunk 9]
```diff
@@ -18,20 +19,44 @@ namespace System.Net.Security.Tests
...
             { SslProtocols.Tls12, s_tls12CipherSuiteLookup },
             { SslProtocols.Tls11, s_tls10And11CipherSuiteLookup },
             { SslProtocols.Tls, s_tls10And11CipherSuiteLookup },
         };
 
+        private static Lazy<bool> s_cipherSuitePolicySupported = new Lazy<bool>(() =>
+        {
+            try
+            {
+                new CipherSuitesPolicy(new TlsCipherSuite[0]);
```
Comment (bartonjs) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamNegotiatedCipherSuiteTest.cs]:
```suggestion
                new CipherSuitesPolicy(Array.Empty<TlsCipherSuite>());
```

File: src/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.OpenSslName.Lookup.Linux.cs (hunk 1 above)
Comment (krwq) [reply src/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.OpenSslName.Lookup.Linux.cs]: I meant the whole file, initially it was doing a bit more but after couple of iterations it ended up being redundant...

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs [hunk 10]
```diff
@@ -64,31 +64,39 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
...
-                        protocols &= ~SslProtocols.Tls13;
-
-                        if (protocols == SslProtocols.None)
-                        {
-                            throw new SslException(
-                                SR.Format(SR.net_ssl_encryptionpolicy_notsupported, policy));
-                        }
+                        // user explicitly asks for TLS 1.3 but their policy is not compatible with TLS 1.3
+                        throw new SslException(
+                            SR.Format(SR.net_ssl_encryptionpolicy_notsupported, policy));
```
[the remark this answers is not in the document]
Comment (krwq) [reply src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: @bartonjs actually this happens when user wants TLS1.3 (protocols include it) but user specified ciphersuites in such a way that TLS1.3 cannot happen (none TLS 1.3 ciphersuites included in a policy) - so user contradicted themselves

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c (hunk 3 above)
Comment (krwq) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: I have ifdefed it in the new iteration (for non-portable it always returns false when API not present)

File: src/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.Linux.cs (hunk 7 above)
Comment (krwq) [reply src/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.Linux.cs]: I didn't know github added this feature

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (bartonjs) [reply src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: >  so user contradicted themselves

Yeah, I know.  The cases where this would come up seem vanishingly small, but the easy one is trying to build something like SslLabs, where you connect, see what ciphersuite got negotiated, remove it from the allow list, and try again.  When the last TLS 1.3 ciphersuite gets removed the next call throws.

The throw just means that we're pushing the knowledge of TLS 1.3 vs prior ciphersuites onto the caller.  Which, given we are encouraging people to just use system defaults (after configuring their system, if they don't like the out-of-the-box answer), seems okay for now.

[the remark this answers is not in the document]
Comment (krwq) [reply src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: I got mixed feelings for that - if user doesn't want to know about TLS 1.3 they should use None for protocols in which case we won't throw.

Perhaps let's leave as is and perhaps change it when someone complains this is not intuitive?

I don't feel strongly either way - allowing this would simplify a code a bit but feels to me that asking for TLS 1.3 and not passing any ciphers which could allow us to do so is a user bug - for scenarios like SslLabs they will have to understand the cipher suites either way...

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c [hunk 11]
```diff
@@ -402,7 +402,11 @@ const char* CryptoNative_GetOpenSslCipherSuiteName(SSL* ssl, int32_t cipherSuite
 
 int32_t CryptoNative_Tls13Supported()
 {
+#ifdef HAVE_OPENSSL_SET_CIPHERSUITES
```
Comment (bartonjs) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: This seems perhaps trickily wrong.

If the native shim is built against OpenSSL 1.1.0 or 1.0.2 in portable mode it will see it doesn't have this function, then drop to return false; even if it was portable-bound to 1.1.1 where it should have returned true.

That is, of course, predicated on me remembering that you didn't force-set HAVE_OPENSSL_SET_CIPHERSUITES when building in portable mode.

Comment (bartonjs) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: Oh, and if it's FALLBACK_FUNCTION now it would return true on portable-bound non-1.1.1, since it'll be asking "does the local_ function exist"

Comment (krwq) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: @bartonjs yes, I've removed that before I had more context on what that was for, will revert it back

**Comment by krwq** [comment]
finally green on the CI. Left:
- apply remaining feedback
- add tests for AllowedCipherSuites property
- double check tests for covering corner cases

**Comment by krwq** [comment]
test this please

**Comment by krwq** [comment]
@safern @dotnet/dnceng how do I get this picked up by CI? this has already been running correctly, then it got green and it doesn't get picked up by CI anymore

**Comment by safern** [comment]
Maybe it is related to it being a draft PR? Can you try closing and reopening?

**Comment by krwq** [comment]
@safern probably related, but seems we shouldn't be doing anything special for draft PRs, right? people use them to test stuff on CI, everything else I can do locally without opening any PR
